File: structural/typescript/bridge_structured/implementations.ts
import { DrawingAPI, Device } from './interfaces';

export class RedCircleAPI implements DrawingAPI {
  drawCircle(x: number, y: number, radius: number): void {
    console.log(`Drawing Circle [color: red, center: (${x}, ${y}), radius: ${radius}]`);
  }
  drawRectangle(x: number, y: number, width: number, height: number): void {
    console.log(`Drawing Rectangle [color: red, top-left: (${x}, ${y}), width: ${width}, height: ${height}]`);
  }
}

export class GreenCircleAPI implements DrawingAPI {
  drawCircle(x: number, y: number, radius: number): void {
    console.log(`Drawing Circle [color: green, center: (${x}, ${y}), radius: ${radius}]`);
  }
  drawRectangle(x: number, y: number, width: number, height: number): void {
    console.log(`Drawing Rectangle [color: green, top-left: (${x}, ${y}), width: ${width}, height: ${height}]`);
  }
}

export class Television implements Device {
  private on: boolean = false;
  private volume: number = 10;
  
  turnOn(): void { this.on = true; console.log('TV is now ON'); }
  turnOff(): void { this.on = false; console.log('TV is now OFF'); }
  setVolume(volume: number): void {
    if (this.on) { this.volume = volume; console.log(`TV volume set to ${volume}`); }
    else { console.log('Cannot set volume: TV is OFF'); }
  }
  getVolume(): number { return this.volume; }
}

export class Radio implements Device {
  private on: boolean = false;
  private volume: number = 5;
  
  turnOn(): void { this.on = true; console.log('Radio is now ON'); }
  turnOff(): void { this.on = false; console.log('Radio is now OFF'); }
  setVolume(volume: number): void {
    if (this.on) { this.volume = volume; console.log(`Radio volume set to ${volume}`); }
    else { console.log('Cannot set volume: Radio is OFF'); }
  }
  getVolume(): number { return this.volume; }
}

File: structural/dart/bridge_structured/implementations.dart
import 'interfaces.dart';

class RedCircleAPI implements DrawingAPI {
  @override
  void drawCircle(double x, double y, double radius) {
    print('Drawing Circle [color: red, center: ($x, $y), radius: $radius]');
  }
  
  @override
  void drawRectangle(double x, double y, double width, double height) {
    print('Drawing Rectangle [color: red, top-left: ($x, $y), width: $width, height: $height]');
  }
}

class GreenCircleAPI implements DrawingAPI {
  @override
  void drawCircle(double x, double y, double radius) {
    print('Drawing Circle [color: green, center: ($x, $y), radius: $radius]');
  }
  
  @override
  void drawRectangle(double x, double y, double width, double height) {
    print('Drawing Rectangle [color: green, top-left: ($x, $y), width: $width, height: $height]');
  }
}

class Television implements Device {
  bool _on = false;
  int _volume = 10;
  
  @override
  void turnOn() {
    _on = true;
    print('TV is now ON');
  }
  
  @override
  void turnOff() {
    _on = false;
    print('TV is now OFF');
  }
  
  @override
  void setVolume(int volume) {
    if (_on) {
      _volume = volume;
      print('TV volume set to $volume');
    } else {
      print('Cannot set volume: TV is OFF');
    }
  }
  
  @override
  int getVolume() {
    return _volume;
  }
}

class Radio implements Device {
  bool _on = false;
  int _volume = 5;
  
  @override
  void turnOn() {
    _on = true;
    print('Radio is now ON');
  }
  
  @override
  void turnOff() {
    _on = false;
    print('Radio is now OFF');
  }
  
  @override
  void setVolume(int volume) {
    if (_on) {
      _volume = volume;
      print('Radio volume set to $volume');
    } else {
      print('Cannot set volume: Radio is OFF');
    }
  }
  
  @override
  int getVolume() {
    return _volume;
  }
}
